### PR TITLE
Add stats::modes functionality

### LINKED ISSUE #8: Returns None when mode of iter has more than one
Is this expected behavior?

```rust
let vec = vec![0, 0, 0, 1, 1, 1, 2];

assert_eq!(stats::mode(vec.into_iter()), None);
```

Seems to me it should return a `Vec<T>` rather than `T`

### PR DESCRIPTION
**Pull request by milesgranger**
Closes #8 

@BurntSushi  Have a look, good sir, let me know what needs to be adjusted. Sure I didn't nail it on the first try. ;)

### CODE REVIEW

File: src/unsorted.rs [hunk 1]
```diff
@@ -22,6 +22,16 @@ pub fn mode<T, I>(it: I) -> Option<T>
     it.collect::<Unsorted<T>>().mode()
 }
 
+/// Compute the modes on a stream of data.
+///
+/// (This has time complexity `O(n)`)
```
Comment (BurntSushi) [inline src/unsorted.rs]: Can you remove the parens here and add a period to the end of the sentence?

File: src/unsorted.rs [hunk 2]
```diff
@@ -71,6 +81,40 @@ fn mode_on_sorted<T, I>(it: I) -> Option<T>
...
+            }
+        } else {
+            values.push(x);
+            modes.push(1);
+            count += 1;
+        }
+    }
+
+
+
```
Comment (BurntSushi) [inline src/unsorted.rs]: Can you remove the extra whitespace here? At most one blank line, please. :-)

File: src/unsorted.rs [hunk 3]
```diff
@@ -71,6 +81,40 @@ fn mode_on_sorted<T, I>(it: I) -> Option<T>
     mode
 }
 
+fn modes_on_sorted<T, I>(it: I) -> Vec<T>
+        where T: PartialOrd, I: Iterator<Item=T> {
+
+    let mut highest_mode = 1_u32;
+    let mut modes: Vec<u32> = Vec::new();
+    let mut values = Vec::new();
```
Comment (BurntSushi) [inline src/unsorted.rs]: Can you use `vec![]` instead of `Vec::new()`?

File: src/unsorted.rs [hunk 4]
```diff
@@ -71,6 +81,40 @@ fn mode_on_sorted<T, I>(it: I) -> Option<T>
...
+            values.push(x);
+            modes.push(1);
+            count += 1;
+        }
+    }
+
+
+
+    modes.into_iter()
+            .zip(values)
```
Comment (BurntSushi) [inline src/unsorted.rs]: Can you use one level of indentation here and for subsequent methods in this chain?

File: src/unsorted.rs [hunk 5]
```diff
@@ -202,4 +255,20 @@ mod test {
         assert_eq!(mode(vec![4.0f64, 3.0, 3.0, 3.0].into_iter()), Some(3.0));
         assert_eq!(mode(vec![1.0f64, 1.0, 2.0, 3.0, 3.0].into_iter()), None);
     }
+
+    #[test]
+    fn modes_stream() {
+        assert_eq!(modes(vec![3usize, 5, 7, 9].into_iter()), vec![]);
+        assert_eq!(modes(vec![3usize, 3, 3, 3].into_iter()), vec![3]);
+        assert_eq!(modes(vec![3usize, 3, 4, 4].into_iter()), vec![3, 4]);
```
Comment (BurntSushi) [inline src/unsorted.rs:272]: I think this is wrong. For a mode to exist, it seems to require one element to be more common than at least one other element. I actually had trouble tracking down a very explicit definition of `mode`, so I went to see what Python does:

```
>>> import statistics
>>> statistics.mode([1])
1
>>> statistics.mode([1, 1])
1
>>> statistics.mode([1, 2])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/statistics.py", line 506, in mode
    'no unique mode; found %d equally common values' % len(table)
statistics.StatisticsError: no unique mode; found 2 equally common values
>>> statistics.mode([])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/statistics.py", line 509, in mode
    raise StatisticsError('no mode for empty data')
statistics.StatisticsError: no mode for empty data
>>> statistics.mode([1, 1, 2])
1
```

We should probably copy Python here. If a mode doesn't exist, then we can just return an empty `Vec`.

**Review by BurntSushi — CHANGES_REQUESTED** [review]
Thanks for working on this! I think this mostly looks good. I left some formatting related comments and one comment about a correctness tweak that I think we should make.

File: src/unsorted.rs (hunk 5 above)
Comment (milesgranger) [reply src/unsorted.rs:272]: If this is the case, then I suppose there is no reason to have the added `modes` method. The entire purpose of adding this was to (albeit, selfishly...) solve a problem I had, but which was _almost_ solved by using `rust-stats`.

Alas, I've added to the documentation along with an example in order to be explicit, that perhaps it doesn't follow the normal practice of what a `mode` is... thus `modes`. 

It's up to you of course; but again, if you don't like how it fits, then in my mind it serves no purpose over the current `stats::mode()`

Comment (BurntSushi) [reply src/unsorted.rs:272]: Well, I guess the mode of `[1, 1, 2, 2, 3]` would be `[1, 2]` for example. Interestingly, Python does require the mode to be unique, but this change seems reasonable to me given the name of the method.

Can you try to find some prior art here on the definition you've implemented? I'm not intrinsically opposed to your semantics, but I also don't want to go and redefine what the common definition of `mode` actually is. If we can find a prominent precedent, then I'm much more comfortable with it. The issue is that I am not a mathematician and therefore don't feel qualified to be unilaterally deciding these things.

Comment (milesgranger) [reply src/unsorted.rs:272]: If [Microsoft Excel can do it](https://support.office.com/en-us/article/mode-mult-function-50fd9464-b2ba-4191-b57a-39446689ae8c) then I think we should be able to do it. ;)

Kidding aside, I don't view it as _redefining_ what a `mode` is, because of the signature itself specifying it's the plural of that: `modes`

And at least one other less fortunate sole [wanted it in Java](https://stackoverflow.com/questions/8858327/finding-multiple-modes-in-an-array). ;)

Comment (BurntSushi) [reply src/unsorted.rs:272]: I feel like we're talking past each other here. On the one hand, I understand the modes of `[1, 1, 2, 2, 3]` to be `[1, 2]`. I'm fine with that behavior. What I don't quite understand is how we should define the semantics of something like `[1, 1, 2, 2]`. From looking at your Excel link, I can't actually tell what it does in the case of `[1, 1, 2, 2]`. Similarly for your SO link. The case of `[1, 1, 2, 2]` isn't covered.

Comment (milesgranger) [reply src/unsorted.rs:272]: I'm sorry. 

According to the current behavior (just added a test for this case now), and what it does in Excel, the behavior is `[1, 1, 2, 2]` -> `[1, 2]`


image

Comment (BurntSushi) [reply src/unsorted.rs:272]: OK. That convinces me. I'm happy to point to Excel. :-)

Comment (BurntSushi) [reply src/unsorted.rs:272]: Can you just add a test case here for the empty vec? After that, I think this is good to go!

Comment (milesgranger) [reply src/unsorted.rs:272]: Great, I've added the test for the empty vec. Let me know if you'd like anything else changed, I really appreciate the help/mentoring. :+1:

**Review by BurntSushi — APPROVED** [review]
LGTM! Thanks so much for working on this and your patience. :-)

**Comment by BurntSushi** [comment]
This PR is on crates.io in `streaming-stats 0.2.2`.
